api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,246 @@
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the IstioCSR.
+	Spec IstioCSRSpec `json:"spec,omitempty"`
+
+	// status is the most recently observed status of the IstioCSR.
+	Status IstioCSRStatus `json:"status,omitempty"`
+}
+
+// IstioCSRSpec is the specification of the desired behavior of the IstioCSR.
+type IstioCSRSpec struct {
+	// istioCSRConfig is for configuring the istio-csr agent behavior.
+	IstioCSRConfig *IstioCSRConfig `json:"istioCSRConfig,omitempty"`
+
+	// controllerConfig is for configuring the controller for setting up
+	// defaults to enable istio-csr agent.
+	ControllerConfig *ControllerConfig `json:"controllerConfig,omitempty"`
+}
+
+// IstioCSRConfig is for configuring the istio-csr agent behavior.
+type IstioCSRConfig struct {
+	// logLevel is for setting verbosity of istio-csr agent logging.
+	// Supported log levels: 1-5.
+	// +kubebuilder:default:=1
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=5
+	// +optional
+	LogLevel int32 `json:"logLevel,omitempty"`
+
+	// logFormat is for specifying the output format of istio-csr agent logging.
+	// Support log formats are text and json.
+	// +kubebuilder:default:=text
+	// +optional
+	LogFormat string `json:"logFormat,omitempty"`
+
+	// certmanager is for configuring cert-manager specifics.
+	// +required
+	CertManager *CertManagerConfig `json:"certmanager,omitempty"`
+
+	// istiodTLSConfig is for configuring istiod certificate specifics.
+	// +required
+	IstiodTLSConfig *IstiodTLSConfig `json:"istiodTLSConfig,omitempty"`
+
+	// server is for configuring the server endpoint used by istio
+	// for obtaining the certificates.
+	// +optional
+	Server *ServerConfig `json:"server,omitempty"`
+
+	// istio is for configuring the istio specifics.
+	// +required
+	Istio *IstioConfig `json:"istio,omitempty"`
+
+	// resources is for defining the resource requirements.
+	// Cannot be updated.
+	// ref: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// affinity is for setting scheduling affinity rules.
+	// ref: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	// tolerations is for setting the pod tolerations.
+	// ref: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
+	// +optional
+	// +listType=atomic
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// nodeSelector is for defining the scheduling criteria using node labels.
+	// ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	// +mapType=atomic
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```

Comment:
I don't think the values of these fields are validated. For the main certmanager operator we've validated these fields before applying to the corresponding operands. We should follow similar approach for the istio-csr controller as well.

Reply:
suggestion incorporated.